Benchmark/towers

IR0NSIGHT opened the pull request:
fixes #14

IR0NSIGHT committed:
towers js files

IR0NSIGHT committed:
pushDisk and buildTower works

phischu commented on src/towers.effekt:
```diff
@@ -0,0 +1,170 @@
+//each list represents one smallerDisk.
+//TODO type alias, rename as "smallerDisk"
+//smallerDisk.next is the smallerDisk below => kellerstack
+type Disk = List[Int]
```
Rename this to `Disks` or `Pile` or `Stack`. A single disk is an integer (its size).

IR0NSIGHT replied:
are we sure we want to deviate from the original?
it would change the semantics quite a lot and probably make it less comparable to the JS implementation of the benchmark.

phischu replied:
Yes, do it. The semantics will stay the same, we are merely using types to enforce preconditions.

phischu commented on src/towers.effekt:
```diff
@@ -0,0 +1,170 @@
+type Disk = List[Int]
+
+
+// def printPiles(piles: Array[Disk]) = {
+//   println(pilesToString(piles))
+// }
+
+// def DiskToString(d: Disk): String = {
+//   "Disk("++helper(d)++")"
+// }
+
+// def helper(d: Disk): String = {
+//   d match {
+//     case Cons(size, next) => ""++show(size)++" " ++ helper(next)
+//     case Nil() => ""
+//   }
+// }
+
+// def assert(myBool: Boolean, s: String) = {
+//   if (not(myBool)) {
+//     panic("assertion failed:"++s)
+//   }
+// }
+
+// def pilesToString(piles: Array[Disk]): String = {
+//   var s = "["
+//   each(0, piles.size()){ i =>
+//     s = s ++ "\n\t" ++ DiskToString(piles.unsafeGet(i))
+//   }
+//   s = s ++ "]"
+//   s
+// }
```
Remove this commented-out code (and the other)

phischu commented on src/towers.effekt:
```diff
@@ -0,0 +1,170 @@
+  innerBenchmarkLoop(1){benchmark}{verifyResult}
+}
+
+def diskSize(d: Disk): Int = {
```
This is `head` on lists. Use this instead.

phischu commented on src/towers.effekt:
```diff
@@ -0,0 +1,170 @@
+  var piles: Array[Disk] = emptyArray();
+  var movesDone: Int = 0;
+
+  def pushDisk(piles: Array[Disk], smallerDisk: Disk, pileIdx: Int): Disk = {
```
`smallerDisk` should be an `Int`. You could introduce an alias `type Disk = Int`.

phischu commented on src/towers.effekt:
```diff
@@ -0,0 +1,170 @@
+    return currentTopDisk 
+  }
+
+  def popDiskFrom(pileIdx: Int): Disk = {
```
This one does not abctract over `piles`, while `pushDisk` does. It should be consistent.

phischu reviewed (changes requested):
This can be much simplified if you define your types as

```
type Disk = Int
type Pile = List[Disk]
```
In this case deviating from the original is warranted.

IR0NSIGHT committed:
run towers 100 times in normal mode

IR0NSIGHT committed:
fix towers benchmark, used list

rename field
add towers benchmark to runner
fix towers benchmark, used list left commented out debug helper stuff in there in case i ever have to touch it again

IR0NSIGHT committed:
javascript towers benchmark runnable

IR0NSIGHT committed:
add towers benchmark to CLI

IR0NSIGHT committed:
fix benchmark import in towers

IR0NSIGHT committed:
Towers: refactor to make Disk, StackOfDisk and pile more understandable

IR0NSIGHT committed:
Towers: tiny refactor

IR0NSIGHT committed:
refactor towers, get rid of obsolete pattern matching
